src/object/srv_ec_aggregate.c

```diff
@@ -0,0 +1,1207 @@
+	else
+		return recx->rx_nr;
+}
+
+/* Iterator call back sub-function for handling data extents.
+ */
+static int
+agg_data_extent(vos_iter_entry_t *entry, struct ec_agg_entry *agg_entry,
+		daos_handle_t ih, unsigned int *acts)
+{
+	struct ec_agg_extent	*extent = NULL;
+	int			 rc = 0;
+
+	if (agg_stripenum(agg_entry, entry->ie_recx.rx_idx) !=
+			agg_entry->ae_cur_stripe.as_stripenum) {
+		if (agg_entry->ae_cur_stripe.as_stripenum != ~0UL)
+			agg_process_stripe(agg_entry);
+		agg_entry->ae_cur_stripe.as_stripenum =
+			agg_stripenum(agg_entry, entry->ie_recx.rx_idx);
+	}
+	D_ALLOC_PTR(extent);
+	if (extent == NULL) {
+		rc = -DER_NOMEM;
+		goto out;
+	}
+	extent->ae_recx = entry->ie_recx;
+	extent->ae_epoch = entry->ie_epoch;
+	agg_entry->ae_rsize = entry->ie_rsize;
+
+	d_list_add_tail(&extent->ae_link,
+			&agg_entry->ae_cur_stripe.as_dextents);
+	agg_entry->ae_cur_stripe.as_extent_cnt++;
+	agg_entry->ae_cur_stripe.as_stripe_fill +=
+		agg_in_stripe(agg_entry, &entry->ie_recx);
+
+	if (extent->ae_epoch > agg_entry->ae_cur_stripe.as_hi_epoch)
+		agg_entry->ae_cur_stripe.as_hi_epoch = extent->ae_epoch;
+
+	D_DEBUG(DB_TRACE, "adding extent %lu,%lu, to stripe  %lu, shard: %u\n",
+		extent->ae_recx.rx_idx, extent->ae_recx.rx_nr,
+		agg_stripenum(agg_entry, extent->ae_recx.rx_idx),
+		agg_entry->ae_oid.id_shard);
+out:
+	return rc;
+}
+
+static int
+agg_akey_post(daos_handle_t ih, vos_iter_entry_t *entry,
+	 struct ec_agg_entry *agg_entry,
+	 unsigned int *acts)
+{
+	int rc = 0;
+
+	if (agg_entry->ae_cur_stripe.as_extent_cnt)
+		agg_process_stripe(agg_entry);
+
+	return rc;
+}
+
+/* Handles each replica extent returned by the RECX iterator.
+ */
+static int
+agg_ev(daos_handle_t ih, vos_iter_entry_t *entry,
+	struct ec_agg_entry *agg_entry, unsigned int *acts)
+{
+	int			rc = 0;
+
+	D_ASSERT(!(entry->ie_recx.rx_idx & PARITY_INDICATOR));
+
+	rc = agg_data_extent(entry, agg_entry, ih, acts);
+
+	return rc;
+}
+
+/* Pre-subtree iteration call back for per-object iterator
+ */
+static int
+agg_iterate_pre_cb(daos_handle_t ih, vos_iter_entry_t *entry,
+	vos_iter_type_t type, vos_iter_param_t *param,
+	void *cb_arg, unsigned int *acts)
+{
+	struct ec_agg_entry	*agg_entry = (struct ec_agg_entry *) cb_arg;
+	int			 rc = 0;
+
+	switch (type) {
+	case VOS_ITER_DKEY:
+		rc = agg_dkey(ih, entry, agg_entry, acts);
+		break;
+	case VOS_ITER_AKEY:
+		rc = agg_akey(ih, entry, agg_entry, acts);
+		break;
+	case VOS_ITER_RECX:
+		rc = agg_ev(ih, entry, agg_entry, acts);
+		break;
+	default:
+		break;
+	}
+
+	return rc;
+}
+
+/* Post iteration call back for per-object iterator
+ */
+static int
+agg_iterate_post_cb(daos_handle_t ih, vos_iter_entry_t *entry,
+	vos_iter_type_t type, vos_iter_param_t *param,
+	void *cb_arg, unsigned int *acts)
+{
+	struct ec_agg_entry	*agg_entry = (struct ec_agg_entry *) cb_arg;
+	int			 rc = 0;
+
+	switch (type) {
+	case VOS_ITER_DKEY:
+		break;
+	case VOS_ITER_AKEY:
+		rc = agg_akey_post(ih, entry, agg_entry, acts);
```

Comment:
I think this is the place do the real work right? the _pre callback attach extents, this function will fetch data, compute parity and write local VOS. 
If my description is true, you might want to add more details comments for this function, because it almost the most important one.

Reply:
The work is done within the the RECX iterator, each time a stripe is completed.  The final stripe for the key is called out of the akey post handler, if there are extents:

static int
agg_akey_post(daos_handle_t ih, vos_iter_entry_t *entry,
         struct ec_agg_entry *agg_entry,
         unsigned int *acts)
{
        int rc = 0;

        if (agg_entry->ae_cur_stripe.as_extent_cnt)
                rc = agg_process_stripe(agg_entry);

        return rc;
}